pkg/entities/record.go

```diff
@@ -99,200 +104,320 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
-			return 0, fmt.Errorf("val argument is not of type uint32")
-		}
-		binary.BigEndian.PutUint32(bytesToAppend, v)
-	case Unsigned64:
-		v, ok := val.(uint64)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type uint64")
-		}
-		binary.BigEndian.PutUint64(bytesToAppend, v)
-	case Signed8:
-		v, ok := val.(int8)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int8")
-		}
-		bytesToAppend[0] = byte(v)
-	case Signed16:
-		v, ok := val.(int16)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int16")
-		}
-		binary.BigEndian.PutUint16(bytesToAppend, uint16(v))
-	case Signed32:
-		v, ok := val.(int32)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int32")
-		}
-		binary.BigEndian.PutUint32(bytesToAppend, uint32(v))
-	case Signed64:
-		v, ok := val.(int64)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int64")
-		}
-		binary.BigEndian.PutUint64(bytesToAppend, uint64(v))
-	case Float32:
-		v, ok := val.(float32)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type float32")
-		}
-		binary.BigEndian.PutUint32(bytesToAppend, math.Float32bits(v))
-	case Float64:
-		v, ok := val.(float64)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type float64")
-		}
-		binary.BigEndian.PutUint64(bytesToAppend, math.Float64bits(v))
-	case Boolean:
-		v, ok := val.(bool)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type bool")
-		}
-		// Following boolean spec from RFC7011
-		if v {
-			bytesToAppend[0] = 1
-		} else {
-			bytesToAppend[0] = 2
-		}
-	case DateTimeSeconds, DateTimeMilliseconds:
-		// We expect time to be given in int64 as unix time type in go
-		v, ok := val.(int64)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type int64")
-		}
-		binary.BigEndian.PutUint64(bytesToAppend, uint64(v))
-		// Currently only supporting seconds and milliseconds
-	case DateTimeMicroseconds, DateTimeNanoseconds:
-		// TODO: RFC 7011 has extra spec for these data types. Need to follow that
-		return 0, fmt.Errorf("This API does not support micro and nano seconds types yet")
-	case MacAddress:
-		// Expects net.Hardware type
-		v, ok := val.(net.HardwareAddr)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type net.HardwareAddr for this element")
-		}
-		for i, b := range v {
-			bytesToAppend[i] = b
-		}
-		//bytesToAppend = append(bytesToAppend, []byte(v)...)
-	case Ipv4Address, Ipv6Address:
-		// Expects net.IP type
-		v, ok := val.(net.IP)
-		if !ok {
-			return 0, fmt.Errorf("val argument is not of type net.IP for this element")
-		}
-		if ipv4Add := v.To4(); ipv4Add != nil {
-			ipv4Int := big.NewInt(0)
-			ipv4Int.SetBytes(ipv4Add)
-			binary.BigEndian.PutUint32(bytesToAppend, uint32(ipv4Int.Uint64()))
-		} else {
-			for i, b := range v {
-				bytesToAppend[i] = b
+	if isDecoding {
+		ieValue := NewInfoElementValue(element, value)
+		d.elements = append(d.elements, ieValue)
+	} else {
+		if element.DataType == String {
```

Comment:
This dataType seems to be handled in convertToType already. Why is there a check again?